src/fs/pfs_filesystem_impl.cpp

```diff
@@ -0,0 +1,569 @@
+        }
+        if (DirExists(path))
+            continue;
+        if (pfs_mkdir(path.c_str(), 0755) < 0) {
+            LOG(WARNING) << "pfs_mkdir " << path
+                         << " failed: "<< strerror(errno);
+            return -errno;
+        }
+    }
+    return 0;
+}
+
+bool PfsFileSystemImpl::DirExists(const string& dirPath) {
+    struct stat path_stat;
+    if (0 == pfs_stat(dirPath.c_str(), &path_stat))
+        return S_ISDIR(path_stat.st_mode);
+    else
+        return false;
+}
+
+bool PfsFileSystemImpl::FileExists(const string& filePath) {
+    struct stat path_stat;
+    if (0 == pfs_stat(filePath.c_str(), &path_stat))
+        return S_ISREG(path_stat.st_mode);
+    else
+        return false;
+}
+
+bool PfsFileSystemImpl::PathExists(const string& path) {
+    return pfs_access(path.c_str(), F_OK) == 0;
+}
+
+int PfsFileSystemImpl::DoRename(const string& oldPath,
+                   const string& newPath,
+                   unsigned int flags) {
+    int rc = pfs_rename2(oldPath.c_str(), newPath.c_str(), flags);
+    if (rc < 0) {
+        LOG(WARNING) << "rename failed: " << strerror(errno)
+                     << ". old path: " << oldPath
+                     << ", new path: " << newPath
+                     << ", flag: " << flags;
+        return -errno;
+    }
+    return 0;
+}
+
+int PfsFileSystemImpl::List(const string& dirPath, vector<std::string>* names) {
+    DIR *dir = pfs_opendir(dirPath.c_str());
+    if (nullptr == dir) {
+        LOG(WARNING) << "pfs_opendir:" << dirPath
+                     << " failed:" << strerror(errno);
+        return -errno;
+    }
+    struct dirent *dirIter;
+    errno = 0;
+    while ((dirIter = pfs_readdir(dir)) != nullptr) {
+        if (strcmp(dirIter->d_name, ".") == 0
+                || strcmp(dirIter->d_name, "..") == 0)
+            continue;
+        names->push_back(dirIter->d_name);
+    }
+    if (errno != 0) {
+        LOG(WARNING) << "readdir failed: " << strerror(errno);
+    }
+    pfs_closedir(dir);
+    return -errno;
+}
+
+DIR* PfsFileSystemImpl::OpenDir(const string& dirPath) {
+    return pfs_opendir(dirPath.c_str());
+}
+
+struct dirent* PfsFileSystemImpl::ReadDir(DIR *dir) {
+    return pfs_readdir(dir);
+}
+
+int PfsFileSystemImpl::CloseDir(DIR *dir) {
+    pfs_closedir(dir);
+    return -errno;
+}
+
+int PfsFileSystemImpl::Read(int fd, char* buf, uint64_t offset, int length) {
+    int remainLength = length;
+    int relativeOffset = 0;
+    int retryTimes = 0;
+    while (remainLength > 0) {
+        int ret = pfs_pread(fd,
+                            buf + relativeOffset,
+                            remainLength,
+                            offset);
+        if (ret == 0) {
+            LOG(WARNING) << "pread returns zero."
+                         << "offset: " << offset
+                         << ", length: " << remainLength;
+            break;
+        }
+        if (ret < 0) {
+            if (errno == EINTR && retryTimes < 3) {
```

Comment:
should you add max retry times in the config file?

Reply:
ext4 filesystem do like this